viam-cartographer/src/slam_service/slam_service.cc

```diff
@@ -579,77 +579,55 @@ void SLAMServiceImpl::GetLatestSampledPointCloudMapString(
-                     samplingFactor;
-    if (skip_count == 0) {
-        skip_count = 1;
-    }
-
     std::string data_buffer;
-
-    // Loop to sample data and reduce resolution. Increments multiplied
-    // by 4 to represent 4 bytes per pixel
-    for (int i = 0; i < size_data; i += skip_count * 4) {
-        // skip pixels that are not in our map(black/past walls)
-        // this check represents [102,102,102]
-        if ((data_vect[i + 0] == defaultCairosEmptyPaintedSlice) &&
-            (data_vect[i + 1] == defaultCairosEmptyPaintedSlice) &&
-            (data_vect[i + 2] == defaultCairosEmptyPaintedSlice))
-            continue;
-
-        // Determine probability based on color pixel
-        int prob = viam::ViamColorToProbability((int)data_vect[i + 2]);
-        if (prob == 0) continue;
-
-        num_points++;
-
-        int pixel_index = i / 4;
-        int pixel_x = pixel_index % width;
-        int pixel_y = pixel_index / width;
-
-        // Convert pixel location to pointcloud point in meters
-        float x_pos = (pixel_x - painted_slices->origin.x()) * kPixelSize;
-        // Y is inverted to match output from getPosition()
-        float y_pos = -(pixel_y - painted_slices->origin.y()) * kPixelSize;
-        // 2D SLAM so Z is set to 0
-        float z_pos = 0;
-
-        // Turn the map point into a vector to perform transformations with.
-        // Current transformation rotates coordinates to match slam service
-        // expectation (XZ plane)
-        Eigen::Vector3d map_point(x_pos, y_pos, z_pos);
-        auto rotated_map_point = pcdRotation * map_point;
-
-        viam::utils::writeFloatToBufferInBytes(data_buffer,
-                                               rotated_map_point.x());
-        viam::utils::writeFloatToBufferInBytes(data_buffer,
-                                               rotated_map_point.y());
-        viam::utils::writeFloatToBufferInBytes(data_buffer,
-                                               rotated_map_point.z());
-        viam::utils::writeIntToBufferInBytes(data_buffer, prob);
+    for (int pixel_y = 0; pixel_y < height; pixel_y++) {
+        for (int pixel_x = 0; pixel_x < width; pixel_x++) {
+            // Get bytes associated with pixel
+            int pixel_index = pixel_x + pixel_y * width;
+            int byte_index = pixel_index * bytesPerPixel;
+            std::vector<unsigned char> pixel_data = {
+                data + byte_index, data + byte_index + bytesPerPixel};
+
+            // Skip pixel if it contains empty data (default color)
+            if (checkIfEmptyPixel(pixel_data)) continue;
+
+            // Determine probability based on color pixel and skip if 0
+            int prob = calculateViamProbabilityFromColorChannel(pixel_data);
+            if (prob == 0) continue;
+
+            // Convert pixel location to pointcloud point in meters
+            float x_pos = (pixel_x - painted_slices->origin.x()) * resolution;
```

Comment:
[note to self] resolution is in meters. so resolution = 3 means you want a pixel every 3 meters.